File: eio2024-ev/2024-12-08-ev/saan/solution/sol_basic.cpp
// O(N^5) dp

#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

typedef long long ll;

int main () {
  int n, K;
  cin >> n >> K;

  vector<int> good (n, 0);
  for (int i = 0; i < K; i++) {
    int x;
    cin >> x;

    good[x - 1] = 1;
  }
  
  vector<vector<vector<ll>>> dp (n + 1, vector<vector<ll>> (n + 1, vector<ll> (n * n + 1, 0)));
  // # placed, # bad, sum
  dp[0][0][0] = 1;

  for (int i = 0; i < n; i++) {
    // try placing i

    for (int j = n - 1; j >= 0; j--) {
      for (int k = n - 1; k >= 0; k--) {
        for (int S = n * n - i; S >= 0; S--) {
          dp[j + 1][k + !good[i]][S + i] += dp[j][k][S];
        }
      }
    }
  }

  if (K * (n - 1) % 2 == 1) {
    cout << -1 << '\n';
    cout << 0 << '\n';
    return 0;
  }
  
  int good_S = K * (n - 1) / 2;
  for (int k = 0; k <= n; k++) {
    if (dp[K][k][good_S] != 0) {
      cout << k << '\n';
      cout << dp[K][k][good_S] << '\n';
      return 0;
    }
  }

  assert(false);
}

File: eio2024-ev/2024-12-08-ev/saan/solution/sol_basic.py
n, K = map(int, input().split())
gs = map(int, input().split())
good = [False] * n
for g in gs:
    good[g - 1] = True

dp = [[[0 for _ in range(n * n + 1)] for __ in range(n + 1)] for ___ in range(n + 1)]
dp[0][0][0] = 1

for i in range(n):
    # try placing i

    for j in range(n - 1, -1, -1):
        for k in range(n - 1, -1, -1):
            for S in range(n * n - i, -1, -1):
                dp[j + 1][k + (not good[i])][S + i] += dp[j][k][S]

if (K * (n - 1)) % 2 == 1:
    print(-1)
    print(0)
else:
    good_S = (K * (n - 1)) // 2
    for k in range(n + 1):
        if dp[K][k][good_S] != 0:
            print(k)
            print(dp[K][k][good_S])
            break
